rep-2007.rst

```diff
@@ -0,0 +1,246 @@
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 28-Jan-2020
+Post-History: 28-Jan-2020
+
+
+Abstract
+========
+
+This REP proposes an extension to the creation of Publishers and Subscriptions in `rclcpp` to make it easier to convert between ROS 2 interfaces and custom C++ data structures.
+
+
+Motivation
+==========
+
+The primary reason for this change is to improve the developer's experience working with ROS 2.
+Currently, developers often write code to convert a ROS interface into another custom data structure for use in their programs.
+This can be trivial, in the case accessing the ``data`` field in ``std_msgs::msg::String``;
+or more involved such when converting OpenCV's ``cv::Map`` to ROS's ``sensor_msgs/msg/Image`` type [1]_.
+Such conversions are additional work for the programmer and are potential sources of errors.
+
+An additional reason for this change is performance.
+This interface will allow us to define methods for serializing directly to the user requested type, and/or using that type in intra-process communication without ever converting it.
+Whereas without this feature, even to send a ``cv::Mat`` from one publisher to a subscription in the same process would require converting it to a ``sensor_msgs::msg::Image`` first and then back again to cv::Mat.
+
+
+Terminology
+===========
+
+.. glossary::
+     Custom type (in code, ``CustomType``)
+       A data structure that *is not* a ROS 2 interface, such as ``std::string`` or ``cv::Mat``.
+     ROS type (in code, ``RosType``)
+       A data structure that *is* a ROS 2 interface, such as ``std_msgs::msg::String`` or ``sensor_msgs::msg::Image``.
+
+
+Specification
+=============
+
+There are different tiers of expressivity in the syntax for type adaptation, including both explicit to implicit forms.
+The explicit forms require the custom type and the ROS type to be specified; 
+while the implicit form makes an assumption about the ROS type based on the custom type.
+The explicit forms may be used to clearly communicate what is happening, which could be useful in examples or tutorials, or may be used depending on the preferences of the developer.
+While the implicit, and most concise, form will primarily be used for its convenience (although it lacks some of the functionality of the explicit forms).
+
+Below are examples of different syntaxes for type adaptation.
+In these examples and using the terminology above, ``std::string`` will be considered the custom type and the ROS interface ``std_msgs::msg::String`` will be considered the ROS type.
+Note that, for every publisher created in the examples below, a ``std::string`` type variable (the custom type) would be adapted to be published on the ROS network as a ``std_msgs::msg::String`` (the ROS type).
+
+.. code-block:: cpp
+
+   // Explicit Form 1: Most explicit
+   using MsgT = rclcpp::AdaptType<std::string>::to<std_msgs::msg::String>;
+   auto pub1 = std::make_shared<rclcpp::Publisher<MsgT>>();
```

Comment:
Maybe instead of creating a publisher that only can publish `std::string` as `std_msgs::msg::String`, we could've the following:

```cpp
auto pub = node->create_publisher<std_msgs::msg::String>(...);
std_msgs::msg::String msg;
msg.data = "publishing a std_msgs::msg::String";
pub->publish(msg);
std::string native_string = "But also a std::string!!!";
pub->publish(native_string);
```

In this case, the `Publisher` template doesn't change, but `publish` has to specializations.
One taking the ros message, the other taking a message that implements `rclcpp::Adapter<native_type, ros_type>`

Reply:
We already have something like:

```c++
auto pub = node->create_publisher<std::string>(...);
std_msgs::msg::String msg;
msg.data = "publishing a std_msgs::msg::String";
pub->publish(msg);
std::string native_string = "But also a std::string!!!";
pub->publish(native_string);
```

in our demo project.

I think that something like this looks like a bug on first glance:

```c++
auto pub = node->create_publisher<std_msgs::msg::String>(...);
std::string native_string = "But also a std::string!!!";
pub->publish(native_string);
```

You could argue that this also looks like a bug:

```c++
auto pub = node->create_publisher<std::string>(...);
std_msgs::msg::String msg;
msg.data = "publishing a std_msgs::msg::String";
pub->publish(msg);
```

But I'd argue that the latter case is far less likely, since you wouldn't need to create the publisher as `std::string` if you didn't intend to publish as `std::string`.

Also, having the adaptor specified in the `Publisher` rather than `Publisher::publish` allows us to prime the intra process system to expect `std::string`. By the time we get to `publish` it might be much more inconvenient.

Reply:
I don't see any of both as a bug really, but also I don't mind strongly if this is supported or not.

> But I'd argue that the latter case is far less likely, since you wouldn't need to create the publisher as std::string if you didn't intend to publish as std::string.

My point of view is that you never have a publisher of a `std::string`, you're always publishing a `std_msgs::msg::String` represented in different ways and `publish` can handle any of the representations (anything that implements rclcpp::TypeAdapter<RepresentationT, MessageT>).

Reply:
I think anyone concerned with the fact that you have `node->create_publisher<std::string>(...)` and that that doesn't reflect what's on the wire can be explicit with `node->create_publisher<rclcpp::TypeAdapter<std::string, std_msgs::msg::String>>(...)` or similar.